File: Python/string_algorithms/08_match_compressed_string.py
import sys
from collections import Counter


def calc_first_occurrence_dict(transformed_string):
    letter_counts = Counter(transformed_string)
    first_occurrence = {}
    ix = 0
    for letter in sorted(letter_counts.keys()):
        first_occurrence[letter] = ix
        ix += letter_counts[letter]
    return first_occurrence


def calc_count_matrix(transformed_string, unique_characters, length):
    count_matrix = {letter: [0] * (length+1) for letter in unique_characters}
    for ix, current_letter in enumerate(transformed_string):
        for char in unique_characters:
            counts = count_matrix[char]
            counts[ix+1] = counts[ix] + 1 if current_letter == char else counts[ix]
    return count_matrix


def letter_occurred_in_last_column_between_pointers(letter, last_column, top_pointer, bottom_pointer):
    return any(letter == char for char in last_column[top_pointer:bottom_pointer+1])


def pattern_matching(pattern, last_column, length, first_occurrences, count_matrix):
    top_pointer = 0
    bottom_pointer = length - 1
    while pattern:
        letter = pattern.pop(-1)
        if not letter_occurred_in_last_column_between_pointers(letter, last_column, top_pointer, bottom_pointer):
            return 0
        top_pointer = first_occurrences[letter] + count_matrix[letter][top_pointer]
        bottom_pointer = first_occurrences[letter] + count_matrix[letter][bottom_pointer+1] - 1
    return bottom_pointer - top_pointer + 1


def improved_BW_pattern_matching(transformed_string, patterns):
    string_length = len(transformed_string)
    first_occurrences = calc_first_occurrence_dict(transformed_string)
    unique_letters = frozenset(first_occurrences.keys())
    count_matrix = calc_count_matrix(transformed_string, unique_letters, string_length)

    pattern_occurrences = []
    for pattern in patterns:
        number_of_matches = pattern_matching(list(pattern), transformed_string, string_length,
                                             first_occurrences, count_matrix)
        pattern_occurrences.append(number_of_matches)
    return pattern_occurrences


def main():
    data = sys.stdin.read().splitlines()
    transformed_string = data[0]
    patterns = data[2].split(" ")
    pattern_matches = improved_BW_pattern_matching(transformed_string, patterns)
    print(" ".join([str(count) for count in pattern_matches]))


if __name__ == "__main__":
    main()

File: Rust/string_algorithms/src/bin/08_match_compressed_string.rs
use std::collections::{HashMap, HashSet};
use std::io;

fn read_line() -> String {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    line
}

fn read_patterns(line: &str) -> Vec<String> {
    line
        .split_whitespace()
        .map(|s| s.to_string())
        .collect()
}

fn calc_first_occurrence_positions(text: &str) -> HashMap<char, i32> {
    let mut letter_counts = HashMap::new();
    for chr in text.chars() {
        letter_counts.entry(chr).and_modify(|cnt| *cnt += 1).or_insert(1);
    }

    let mut chars = letter_counts.keys().collect::<Vec<_>>();
    chars.sort();

    let mut first_occurrence = HashMap::new();
    let mut ix = 0;
    for &chr in chars {
        first_occurrence.insert(chr, ix);
        ix += letter_counts.get(&chr).unwrap();
    }

    first_occurrence
}

fn calc_count_matrix(text: &str, unique_characters: &HashSet<char>) -> HashMap<char, Vec<i32>> {
    let n = text.len();
    let mut count_matrix = HashMap::from_iter(unique_characters.iter().map(|&c| (c, vec![0; n + 1])));
    for (ix, current_chr) in text.chars().enumerate() {
        for chr in unique_characters {
            let counts = count_matrix.get_mut(chr).unwrap();
            counts[ix + 1] = if current_chr == *chr { counts[ix] + 1 } else { counts[ix] };
        }
    }
    count_matrix
}

fn pattern_matching(
    pattern: &str,
    last_column: &[char],
    first_occurrences: &HashMap<char, i32>,
    count_matrix: &HashMap<char, Vec<i32>>,
) -> usize {
    let mut top_pointer = 0;
    let mut bottom_pointer = last_column.len() - 1;
    for chr in pattern.chars().rev() {
        if !last_column[top_pointer..(bottom_pointer+1)].iter().any(|&c| c == chr) {
            return 0
        }
        let char_occurrence = first_occurrences.get(&chr).unwrap();
        let char_counter = count_matrix.get(&chr).unwrap();
        top_pointer = (char_occurrence + char_counter[top_pointer]) as usize;
        bottom_pointer = (char_occurrence + char_counter[bottom_pointer + 1] - 1) as usize;
    }

    bottom_pointer + 1 - top_pointer
}

fn improved_burrows_wheeler_pattern_matching(transformed_text: &str, patterns: &Vec<String>) -> Vec<usize> {
    let first_occurrences = calc_first_occurrence_positions(transformed_text);
    let unique_letters: HashSet<char> = HashSet::from_iter(first_occurrences.clone().into_keys());
    let count_matrix = calc_count_matrix(transformed_text, &unique_letters);

    let last_column = transformed_text.chars().collect::<Vec<_>>();
    let mut pattern_occurrences = Vec::new();
    for pattern in patterns {
        let nr_matches = pattern_matching(pattern, &last_column, &first_occurrences, &count_matrix);
        pattern_occurrences.push(nr_matches);
    }

    pattern_occurrences
}

fn main() {
    let transformed_text = read_line().trim().to_string();
    let _ = read_line();
    let patterns = read_patterns(&read_line());
    let result = improved_burrows_wheeler_pattern_matching(&transformed_text, &patterns);
    println!("{}", result.iter().map(|n| n.to_string()).collect::<Vec<_>>().join(" "));
}
